fix for service auth behavior

birojnayak opened the pull request.

mconnew commented on src/CoreWCF.Primitives/src/CoreWCF/Description/ServiceAuthorizationBehavior.cs:
```diff
@@ -112,8 +112,7 @@ public bool ImpersonateCallerForAllOperations
             set
             {
-                ThrowIfImmutable();
-                impersonateCallerForAllOperations = value;
```
Can you leave the old code commented like you did below so it's obvious what should be here. Does impersonation not work? I put some effort in to making impersonation work with different api's now.

birojnayak replied:
Will write a test to see if it works

mconnew commented on src/CoreWCF.Primitives/src/CoreWCF/IdentityModel/Policy/AuthorizationContext.cs:
```diff
@@ -14,8 +14,8 @@ internal abstract class AuthorizationContext : IAuthorizationComponent
-            throw new PlatformNotSupportedException();
+            return SecurityUtils.CreateDefaultAuthorizationContext(authorizationPolicies);
+            //throw new PlatformNotSupportedException();
         }
```
No need to leave the throw behind in a comment.

mconnew commented on src/CoreWCF.Primitives/src/CoreWCF/IdentityModel/Policy/DefaultAuthorizationContext.cs:
```diff
@@ -0,0 +1,68 @@
+            {
+                /* commenting 
+                if (LocalAppContextSwitches.EnableCachedEmptyDefaultAuthorizationContext)
+                {
```
You can remove this commented code. This was put there as a back compat switch to revert the behavior of a fix. The problem was custom code was incorrectly modifying the empty default authotization context which resulted in leaking auth info from one call to another. The fix was to return a new one each time. My preferred fix at the time was to make it immutable and throw if trying to modify but there was concern that would be an even bigger breaking change. What are your thoughts? Should we make this an immutable singleton or allocate a new one each time?

mconnew commented on src/CoreWCF.Http/tests/ServiceAuthBehaviorTest.cs:
```diff
@@ -0,0 +1,93 @@
+                services.AddServiceModelServices();
+                var authBehavior = new ServiceAuthorizationBehavior();
+                services.AddSingleton<IServiceBehavior>(authBehavior);
+                services.AddSingleton(authBehavior);
```
Should we add this by default in `services.AddServiceModelServices()`? What do you think about enabling people to do `services.AddSingleton<ServiceAuthorizationManager, MyTestServiceAuthorizationManager>()` and have it auto-wire in to the ServiceAuthorizationBehavior?

mconnew commented on src/CoreWCF.Http/tests/ClientContract/IEchoService.cs:
```diff
@@ -30,5 +30,8 @@ public interface IEchoService
 
+        [OperationContract(Name = "EchoStringToFail", Action = Constants.OPERATION_BASE + "EchoStringToFail",
+     ReplyAction = Constants.OPERATION_BASE + "EchoStringResponse")]
+        string EchoToFail(string echo);
```
ReplyAction should be EchoStringToFailResponse. This works on HTTP, but would likely break on NetTcp because it's duplex.

birojnayak commented:
Updated review comments and added more functionalities. NetTCP for linux needs to be fixed (when security is transport), so made few tests Windows only. Once I fix the NetTCP for Linux will enable tests for Linux env.

mconnew commented on src/CoreWCF.NetTcp/tests/Services/TestService.cs:
```diff
@@ -50,5 +54,34 @@ public TestMessage TestMessageContract(TestMessage testMessage)
+            IPrincipal principal = Thread.CurrentPrincipal;
+            if (echo.Contains(PrincipalPermissionMode.UseWindowsGroups.ToString()))
+            {
+                if (principal.Identity !=null
```
Nit: Code formatting, missing space after =

mconnew commented on src/CoreWCF.NetTcp/tests/Services/TestService.cs:
```diff
@@ -50,5 +54,34 @@ public TestMessage TestMessageContract(TestMessage testMessage)
+        {
+            IPrincipal principal = Thread.CurrentPrincipal;
+            if (echo.Contains(PrincipalPermissionMode.UseWindowsGroups.ToString()))
+            {
```
would nameof(PrincipalPermissionMode.UseWindowsGroups) work here?

mconnew commented on src/CoreWCF.NetTcp/tests/Services/TestService.cs:
```diff
@@ -50,5 +54,34 @@ public TestMessage TestMessageContract(TestMessage testMessage)
+            {
+                if (principal.Identity !=null
+                     && principal.Identity.GetType().Equals(typeof(WindowsIdentity)))
+                    return echo;
```
It might be better to use `typeof(WindowsIdentity).IsAssignableFrom(principal.Identity?.GetType())`. That way we don't break if NegotiateStream starts returning a type derived from WindowsIdentity and you can remove the null check.

FreeAndNil replied:
Why not "is WindowsIdentity"?

mconnew replied:
That works too 😄

mconnew commented on src/CoreWCF.Primitives/src/CoreWCF/Dispatcher/DispatchOperationRuntime.cs:
```diff
@@ -403,7 +403,7 @@ internal async Task<MessageRpc> InvokeAsync(MessageRpc rpc)
 
-                    if (parent.SecurityImpersonation?.IsSecurityContextImpersonationRequired(rpc) ?? false)
+                    if (parent.SecurityImpersonation !=null)
                     {
```
Nit, formatting

mconnew commented on src/CoreWCF.Primitives/src/CoreWCF/IdentityModel/Policy/DefaultAuthorizationContext.cs:
```diff
@@ -0,0 +1,61 @@
+            get
+            {
+               if(empty == null)
+                    empty = new DefaultAuthorizationContext(new DefaultEvaluationContext());
```
This code is still using a cached single DefaultAuthorizationContext. We need to either return an immutable context or not cache a single instance. As this code currently is written, a badly written authorization manager can leak claims between calls.

mconnew commented:
Just a couple of minor issues to be addressed, then I'll merge.

mconnew commented on src/CoreWCF.Primitives/src/CoreWCF/IdentityModel/Policy/DefaultAuthorizationContext.cs:
```diff
@@ -0,0 +1,57 @@
+            get
+            {
+               return new DefaultAuthorizationContext(new DefaultEvaluationContext()); ;
+            }
```
Nit: extra colon at end of line
